mir-ci/mir_ci/tests/robot/platforms/vnc/Vnc.py

```diff
@@ -0,0 +1,211 @@
+    MIDDLE = 2
+
+
+@library(scope="TEST")
+class Vnc:
+    """
+    A Robot Framework library for interacting with HIDs through VNC.
+    """
+
+    ROBOT_LISTENER_API_VERSION = 3
+    TOLERANCE = 0.8
+
+    def __init__(self, host="localhost", port=5900, username=None, password=None) -> None:
+        self.ROBOT_LIBRARY_LISTENER = self
+        self._client: asyncvnc.Client = None
+        self._host = host
+        self._port = port
+        self._username = username
+        self._password = password
+        self._pointer_position: Optional[Tuple[int, int]] = None
+        self._rpa_images = Images()
+
+    @keyword
+    async def match(self, template: str, timeout: int = 5) -> List[dict]:
+        """
+        Grab screenshots and compare until there's a match with the provided
+        template.
+
+        :param template: path to an image file to be used as template
+        :param timeout: timeout in seconds
+        :return: list of matched regions
+        :raises ImageNotFoundError: if no match is found within the timeout
+        """
+        await self.connect()
+        regions = []
+        end_time = time.time() + float(timeout)
+        screenshot = None
+        while time.time() <= end_time:
+            screenshot = Image.fromarray(await asyncio.wait_for(self._client.screenshot(), timeout))
+            try:
+                regions = self._rpa_images.find_template_in_image(
+                    screenshot,
+                    template,
+                    tolerance=self.TOLERANCE,
+                )
+            except (RuntimeError, ValueError, ImageNotFoundError):
+                continue
+            else:
+                break
+        else:
+            if screenshot:
+                self._log_failed_match(screenshot, template)
+            raise ImageNotFoundError
+
+        return [
+            {
+                "left": region.left,
+                "top": region.top,
+                "right": region.right,
+                "bottom": region.bottom,
+            }
+            for region in regions
+        ]
+
+    @keyword
+    async def type_string(self, text: str):
+        await self.connect()
+        self._client.keyboard.write(text)
+
+    @keyword
+    async def move_pointer_to_absolute(self, x: int, y: int) -> None:
+        await self.connect()
+        self._client.mouse.move(x, y)
+        self._pointer_position = (x, y)
+
+    @keyword
+    async def move_pointer_to_proportional(self, x: float, y: float) -> tuple[int, int]:
+        await self.connect()
+        absolute_position = self.get_absolute_from_proportional(x, y)
+        self.move_pointer_to_absolute(absolute_position[0], absolute_position[1])
+        return absolute_position
+
+    @keyword
+    async def walk_pointer_to_absolute(self, x: int, y: int, step_distance: int, delay: float) -> None:
+        await self.connect()
+        assert self._pointer_position is not None, "Cannot walk without moving the pointer at least once"
```

Comment:
Kind of a funky precondition in my opinion. Is there any way to just "get" the pointer position at connection time? If it's a bunch of work, we can leave this out :)

Reply:
This is copied from WaylandHID. We can't query the pointer position, but I think we can default it to (0,0), like asyncvnc does.

Reply:
Done.

Reply:
Should we make (0, 0) the default on other platforms?

Reply:
I've added this behavior in the last commit. Setting the default will only happen if we walk the pointer without moving it first. It seems better than issuing an error.